endpoints/apiconfig.go

```diff
@@ -744,3 +744,46 @@ func parseValue(s string, k reflect.Kind) (interface{}, error) {
 
 	return nil, fmt.Errorf("parseValue: Invalid kind %#v value=%q", k, s)
 }
+
+func validateRequest(r interface{}) error {
+	v := reflect.ValueOf(r)
+	if v.Kind() != reflect.Ptr {
+		return fmt.Errorf("%T is not a pointer", r)
+	}
+	v = reflect.Indirect(v)
+	if v.Kind() != reflect.Struct {
+		return fmt.Errorf("%T is not a pointer to a struct", r)
+	}
+
+	t := v.Type()
+	for i := 0; i < v.NumField(); i++ {
+		if err := validateField(v.Field(i), t.Field(i)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func validateField(v reflect.Value, t reflect.StructField) error {
+	tag, err := parseTag(t.Tag)
+	if err != nil {
+		return fmt.Errorf("parse tag: %v", err)
+	}
+	zero := reflect.Zero(v.Type()).Interface()
```

Comment:
Now that you've inlined all this, how about ditching earlier on non-zero?

if zero := reflect.Zero(v.Type()).Interface(); zero != v.Interface() {
  return nil
}

if tag.required { // etc. }}

if tag.default == "" {
  return nil
}

etc.